File: resources/Examples/compact_osw/measures/EpModelMultizoneVAV/measure.rb
class EpModelMultizoneVAV < OpenStudio::Measure::ModelMeasure
  SYSTEM_PREFIX = 'EpModel Multizone VAV'

  def name
    return SYSTEM_PREFIX
  end

  def description
    return 'Replace existing air loops with one multizone VAV system served by central hot water and chilled water plants.'
  end

  def modeler_description
    return 'Finds zones served by existing AirLoopHVAC objects, removes those air loops, and reconnects the zones to a new multizone air system with water heating, water cooling, a boiler hot water loop, and an air-cooled chiller chilled water loop.'
  end

  def arguments(model)
    return OpenStudio::Measure::OSArgumentVector.new
  end

  def build_hot_water_loop(model)
    hot_water_schedule = OpenStudio::EPModel::ScheduleConstant.new(model)
    hot_water_schedule.setName("#{SYSTEM_PREFIX} Hot Water Setpoint")
    hot_water_schedule.setValue(82.0)

    loop = OpenStudio::EPModel::PlantLoop.new(model)
    loop.sizingPlant.setLoopType('Heating')
    loop.sizingPlant.setDesignLoopExitTemperature(82.0)
    loop.sizingPlant.setLoopDesignTemperatureDifference(11.0)

    pump = OpenStudio::EPModel::PumpVariableSpeed.new(model)
    pump.setName("#{SYSTEM_PREFIX} Hot Water Pump")
    pump.addToNode(loop.supplyInletNode)

    boiler = OpenStudio::EPModel::BoilerHotWater.new(model)
    boiler.setName("#{SYSTEM_PREFIX} Hot Water Boiler")
    boiler.setFuelType('NaturalGas')
    boiler.autosizeNominalCapacity
    boiler.autosizeDesignWaterFlowRate
    loop.addSupplyBranchForComponent(boiler)

    heating_operation = OpenStudio::EPModel::PlantEquipmentOperationHeatingLoad.new(model)
    heating_operation.setName("#{SYSTEM_PREFIX} Hot Water Heating Operation Scheme")
    heating_operation.addEquipment(boiler)
    loop.setPlantEquipmentOperationHeatingLoad(heating_operation)

    supply_bypass = OpenStudio::EPModel::PipeAdiabatic.new(model)
    supply_bypass.setName("#{SYSTEM_PREFIX} Hot Water Supply Bypass")
    loop.addSupplyBranchForComponent(supply_bypass)

    supply_outlet_pipe = OpenStudio::EPModel::PipeAdiabatic.new(model)
    supply_outlet_pipe.setName("#{SYSTEM_PREFIX} Hot Water Supply Outlet Pipe")
    supply_outlet_pipe.addToNode(loop.supplyOutletNode)

    demand_inlet_pipe = OpenStudio::EPModel::PipeAdiabatic.new(model)
    demand_inlet_pipe.setName("#{SYSTEM_PREFIX} Hot Water Demand Inlet Pipe")
    demand_inlet_pipe.addToNode(loop.demandInletNode)

    demand_bypass = OpenStudio::EPModel::PipeAdiabatic.new(model)
    demand_bypass.setName("#{SYSTEM_PREFIX} Hot Water Demand Bypass")
    loop.addDemandBranchForComponent(demand_bypass)

    demand_outlet_pipe = OpenStudio::EPModel::PipeAdiabatic.new(model)
    demand_outlet_pipe.setName("#{SYSTEM_PREFIX} Hot Water Demand Outlet Pipe")
    demand_outlet_pipe.addToNode(loop.demandOutletNode)

    spm = OpenStudio::EPModel::SetpointManagerScheduled.new(model)
    spm.setName("#{SYSTEM_PREFIX} Hot Water Setpoint Manager")
    spm.setSchedule(hot_water_schedule)
    spm.addToNode(loop.supplyOutletNode)
    loop.setLoopTemperatureSetpointNode(loop.supplyOutletNode) if loop.respond_to?(:setLoopTemperatureSetpointNode)

    loop
  end

  def build_chilled_water_loop(model)
    chilled_water_schedule = OpenStudio::EPModel::ScheduleConstant.new(model)
    chilled_water_schedule.setName("#{SYSTEM_PREFIX} Chilled Water Setpoint")
    chilled_water_schedule.setValue(7.22)

    loop = OpenStudio::EPModel::PlantLoop.new(model)
    loop.sizingPlant.setLoopType('Cooling')
    loop.sizingPlant.setDesignLoopExitTemperature(7.22)
    loop.sizingPlant.setLoopDesignTemperatureDifference(6.67)

    pump = OpenStudio::EPModel::PumpVariableSpeed.new(model)
    pump.setName("#{SYSTEM_PREFIX} Chilled Water Pump")
    pump.addToNode(loop.supplyInletNode)

    chiller = OpenStudio::EPModel::ChillerElectricEIR.new(model)
    chiller.setName("#{SYSTEM_PREFIX} Air-Cooled Chiller")
    loop.addSupplyBranchForComponent(chiller)

    cooling_operation = OpenStudio::EPModel::PlantEquipmentOperationCoolingLoad.new(model)
    cooling_operation.setName("#{SYSTEM_PREFIX} Chilled Water Cooling Operation Scheme")
    cooling_operation.addEquipment(chiller)
    loop.setPlantEquipmentOperationCoolingLoad(cooling_operation)

    supply_bypass = OpenStudio::EPModel::PipeAdiabatic.new(model)
    supply_bypass.setName("#{SYSTEM_PREFIX} Chilled Water Supply Bypass")
    loop.addSupplyBranchForComponent(supply_bypass)

    supply_outlet_pipe = OpenStudio::EPModel::PipeAdiabatic.new(model)
    supply_outlet_pipe.setName("#{SYSTEM_PREFIX} Chilled Water Supply Outlet Pipe")
    supply_outlet_pipe.addToNode(loop.supplyOutletNode)

    demand_inlet_pipe = OpenStudio::EPModel::PipeAdiabatic.new(model)
    demand_inlet_pipe.setName("#{SYSTEM_PREFIX} Chilled Water Demand Inlet Pipe")
    demand_inlet_pipe.addToNode(loop.demandInletNode)

    demand_bypass = OpenStudio::EPModel::PipeAdiabatic.new(model)
    demand_bypass.setName("#{SYSTEM_PREFIX} Chilled Water Demand Bypass")
    loop.addDemandBranchForComponent(demand_bypass)

    demand_outlet_pipe = OpenStudio::EPModel::PipeAdiabatic.new(model)
    demand_outlet_pipe.setName("#{SYSTEM_PREFIX} Chilled Water Demand Outlet Pipe")
    demand_outlet_pipe.addToNode(loop.demandOutletNode)

    spm = OpenStudio::EPModel::SetpointManagerScheduled.new(model)
    spm.setName("#{SYSTEM_PREFIX} Chilled Water Setpoint Manager")
    spm.setSchedule(chilled_water_schedule)
    spm.addToNode(loop.supplyOutletNode)
    loop.setLoopTemperatureSetpointNode(loop.supplyOutletNode) if loop.respond_to?(:setLoopTemperatureSetpointNode)

    loop
  end

  def build_multizone_air_loop(model, hot_water_loop, chilled_water_loop)
    deck_schedule = OpenStudio::EPModel::ScheduleConstant.new(model)
    deck_schedule.setName("#{SYSTEM_PREFIX} Deck Air Setpoint")
    deck_schedule.setValue(12.8)

    air_loop = OpenStudio::EPModel::AirLoopHVAC.new(model)
    air_loop.setName("#{SYSTEM_PREFIX} System")

    sizing_system = air_loop.sizingSystem
    sizing_system.setCentralCoolingDesignSupplyAirTemperature(12.8)
    sizing_system.setCentralHeatingDesignSupplyAirTemperature(16.7)
    sizing_system.setTypeofLoadtoSizeOn('Sensible') if sizing_system.respond_to?(:setTypeofLoadtoSizeOn)

    outdoor_air_system = OpenStudio::EPModel::AirLoopHVACOutdoorAirSystem.new(model)
    outdoor_air_system.setName("#{SYSTEM_PREFIX} Outdoor Air System")
    controller = outdoor_air_system.getControllerOutdoorAir
    controller.setName("#{SYSTEM_PREFIX} Outdoor Air Controller")

    cooling_coil = OpenStudio::EPModel::CoilCoolingWater.new(model)
    cooling_coil.setName("#{SYSTEM_PREFIX} Central Cooling Coil")
    cooling_coil.autosizeDesignWaterFlowRate
    cooling_coil.autosizeDesignAirFlowRate

    heating_coil = OpenStudio::EPModel::CoilHeatingWater.new(model)
    heating_coil.setName("#{SYSTEM_PREFIX} Central Heating Coil")

    fan = OpenStudio::EPModel::FanVariableVolume.new(model)
    fan.setName("#{SYSTEM_PREFIX} Supply Fan")

    supply_outlet_node = air_loop.supplyOutletNode
    outdoor_air_system.addToNode(supply_outlet_node)
    cooling_coil.addToNode(supply_outlet_node)
    heating_coil.addToNode(supply_outlet_node)
    fan.addToNode(supply_outlet_node)

    chilled_water_loop.addDemandBranchForComponent(cooling_coil)
    hot_water_loop.addDemandBranchForComponent(heating_coil)
    spm = OpenStudio::EPModel::SetpointManagerScheduled.new(model)
    spm.setName("#{SYSTEM_PREFIX} Deck Air Setpoint Manager")
    spm.setSchedule(deck_schedule)
    spm.addToNode(supply_outlet_node)

    cooling_coil_outlet_spm = OpenStudio::EPModel::SetpointManagerScheduled.new(model)
    cooling_coil_outlet_spm.setName("#{SYSTEM_PREFIX} Cooling Coil Outlet Setpoint Manager")
    cooling_coil_outlet_spm.setSchedule(deck_schedule)
    cooling_coil_outlet_spm.addToNode(heating_coil.airInletModelObject.get.to_Node.get)

    heating_coil_outlet_spm = OpenStudio::EPModel::SetpointManagerScheduled.new(model)
    heating_coil_outlet_spm.setName("#{SYSTEM_PREFIX} Heating Coil Outlet Setpoint Manager")
    heating_coil_outlet_spm.setSchedule(deck_schedule)
    heating_coil_outlet_spm.addToNode(fan.inletModelObject.get.to_Node.get)

    return air_loop
  end

  def run(model, runner, user_arguments)
    super(model, runner, user_arguments)

    if !runner.validateUserArguments(arguments(model), user_arguments)
      return false
    end

    zones = []
    air_loops = model.getAirLoopHVACs()
    air_loops.each do |loop|
      zones.concat(loop.thermalZones())
      loop.remove()
    end

    hot_water_loop = build_hot_water_loop(model)
    chilled_water_loop = build_chilled_water_loop(model)
    hot_water_loop.setName("#{SYSTEM_PREFIX} Hot Water Loop")
    chilled_water_loop.setName("#{SYSTEM_PREFIX} Chilled Water Loop")
    air_loop = build_multizone_air_loop(model, hot_water_loop, chilled_water_loop)

    zones.each do |zone|
      reheat_coil = OpenStudio::EPModel::CoilHeatingWater.new(model)
      reheat_coil.setName("#{zone.nameString} Reheat Coil")
      hot_water_loop.addDemandBranchForComponent(reheat_coil)

      terminal = OpenStudio::EPModel::AirTerminalSingleDuctVAVReheat.new(model)
      terminal.setName("#{zone.nameString} VAV Reheat Terminal")
      terminal.setReheatCoil(reheat_coil)
      air_loop.addBranchForZone(zone, terminal)
    end

    runner.registerFinalCondition("Removed #{air_loops.size} AirLoopHVAC object(s), created one multizone VAV air system, and reattached #{zones.size} zone(s) with boiler hot water and chilled water plant service.")
    return true
  end
end

EpModelMultizoneVAV.new.registerWithApplication

File: resources/Examples/epmodel/measures/EpModelAddMultizoneVAVSystem/measure.py
import openstudio


class EpModelAddMultizoneVAVSystem(openstudio.measure.ModelMeasure):
    SYSTEM_PREFIX = "EPModel Multizone VAV"

    def name(self):
        return "EPModel Add Multizone VAV System"

    def description(self):
        return (
            "Serve every thermostatically controlled zone with one multizone VAV system and central hot-water and "
            "chilled-water plants."
        )

    def modeler_description(self):
        return (
            "Consumes the same HVAC-ready contract as the other addition measures, then builds a variable-volume air "
            "system with hot-water reheat, a boiler hot-water loop, and an air-cooled-chiller chilled-water loop."
        )

    def arguments(self, model=None):
        return openstudio.measure.OSArgumentVector()

    @staticmethod
    def require_condition(runner, condition, message):
        if condition:
            return True
        runner.registerError(message)
        return False

    def build_plant_loop(self, model, runner, zone, heating, zone_count):
        kind = "Hot Water" if heating else "Chilled Water"
        temperature = 82.0 if heating else 7.22
        delta = 11.0 if heating else 6.67
        flow_rate = (0.00020 if heating else 0.00042) * zone_count * 1.25

        schedule = openstudio.epmodel.ScheduleConstant(model)
        schedule.setName(f"{self.SYSTEM_PREFIX} {kind} Setpoint")
        if not self.require_condition(runner, schedule.setValue(temperature), f"Could not set the {kind.lower()} setpoint."):
            return None

        loop = openstudio.epmodel.PlantLoop(model)
        sizing = loop.sizingPlant()
        if not self.require_condition(
            runner,
            all(
                [
                    sizing.setLoopType("Heating" if heating else "Cooling"),
                    sizing.setDesignLoopExitTemperature(temperature),
                    sizing.setLoopDesignTemperatureDifference(delta),
                    loop.setMaximumLoopFlowRate(flow_rate),
                ]
            ),
            f"Could not configure the {kind.lower()} loop sizing.",
        ):
            return None

        pump = openstudio.epmodel.PumpVariableSpeed(model)
        pump.setName(f"{self.SYSTEM_PREFIX} {kind} Pump")
        pump.autosizeRatedPowerConsumption()
        if not self.require_condition(
            runner,
            all(
                [
                    pump.setRatedFlowRate(flow_rate),
                    pump.setRatedPumpHead(120_000.0),
                    pump.setMotorEfficiency(0.9),
                    pump.setZone(zone),
                    pump.addToNode(loop.supplyInletNode()),
                ]
            ),
            f"Could not configure and connect the {kind.lower()} pump.",
        ):
            return None

        if heating:
            equipment = openstudio.epmodel.BoilerHotWater(model)
            equipment.setName(f"{self.SYSTEM_PREFIX} Hot Water Boiler")
            if not self.require_condition(runner, equipment.setFuelType("NaturalGas"), "Could not set the boiler fuel type."):
                return None
            equipment.autosizeNominalCapacity()
            equipment.autosizeDesignWaterFlowRate()
            operation = openstudio.epmodel.PlantEquipmentOperationHeatingLoad(model)
            operation.setName(f"{self.SYSTEM_PREFIX} Hot Water Heating Operation Scheme")
            operation_is_set = operation.addEquipment(equipment) and loop.setPlantEquipmentOperationHeatingLoad(operation)
        else:
            equipment = openstudio.epmodel.ChillerElectricEIR(model)
            equipment.setName(f"{self.SYSTEM_PREFIX} Air-Cooled Chiller")
            operation = openstudio.epmodel.PlantEquipmentOperationCoolingLoad(model)
            operation.setName(f"{self.SYSTEM_PREFIX} Chilled Water Cooling Operation Scheme")
            operation_is_set = operation.addEquipment(equipment) and loop.setPlantEquipmentOperationCoolingLoad(operation)

        if not self.require_condition(
            runner,
            operation_is_set and loop.addSupplyBranchForComponent(equipment),
            f"Could not connect the {kind.lower()} equipment and operation scheme.",
        ):
            return None

        for label, placement in [
            ("Supply Bypass", "supply_branch"),
            ("Supply Outlet Pipe", "supply_outlet"),
            ("Demand Inlet Pipe", "demand_inlet"),
            ("Demand Bypass", "demand_branch"),
            ("Demand Outlet Pipe", "demand_outlet"),
        ]:
            pipe = openstudio.epmodel.PipeAdiabatic(model)
            pipe.setName(f"{self.SYSTEM_PREFIX} {kind} {label}")
            if placement == "supply_branch":
                connected = loop.addSupplyBranchForComponent(pipe)
            elif placement == "supply_outlet":
                connected = pipe.addToNode(loop.supplyOutletNode())
            elif placement == "demand_inlet":
                connected = pipe.addToNode(loop.demandInletNode())
            elif placement == "demand_branch":
                connected = loop.addDemandBranchForComponent(pipe)
            else:
                connected = pipe.addToNode(loop.demandOutletNode())
            if not self.require_condition(runner, connected, f"Could not connect the {kind.lower()} {label.lower()}."):
                return None

        spm = openstudio.epmodel.SetpointManagerScheduled(model)
        spm.setName(f"{self.SYSTEM_PREFIX} {kind} Setpoint Manager")
        if not self.require_condition(
            runner,
            spm.setSchedule(schedule)
            and spm.addToNode(loop.supplyOutletNode())
            and loop.setLoopTemperatureSetpointNode(loop.supplyOutletNode()),
            f"Could not configure the {kind.lower()} setpoint manager.",
        ):
            return None

        return loop

    def build_multizone_air_loop(self, model, runner, hot_water_loop, chilled_water_loop):
        deck_schedule = openstudio.epmodel.ScheduleConstant(model)
        deck_schedule.setName(f"{self.SYSTEM_PREFIX} Deck Air Setpoint")
        if not self.require_condition(runner, deck_schedule.setValue(12.8), "Could not set the VAV deck-air setpoint."):
            return None

        air_loop = openstudio.epmodel.AirLoopHVAC(model)
        air_loop.setName(f"{self.SYSTEM_PREFIX} System")
        sizing = air_loop.sizingSystem()
        sizing.setCentralCoolingDesignSupplyAirTemperature(12.8)
        sizing.setCentralHeatingDesignSupplyAirTemperature(16.7)
        if hasattr(sizing, "setTypeofLoadtoSizeOn"):
            sizing.setTypeofLoadtoSizeOn("Sensible")

        outdoor_air_system = openstudio.epmodel.AirLoopHVACOutdoorAirSystem(model)
        outdoor_air_system.setName(f"{self.SYSTEM_PREFIX} Outdoor Air System")
        outdoor_air_system.getControllerOutdoorAir().setName(f"{self.SYSTEM_PREFIX} Outdoor Air Controller")

        cooling_coil = openstudio.epmodel.CoilCoolingWater(model)
        cooling_coil.setName(f"{self.SYSTEM_PREFIX} Central Cooling Coil")
        cooling_coil.autosizeDesignWaterFlowRate()
        cooling_coil.autosizeDesignAirFlowRate()

        heating_coil = openstudio.epmodel.CoilHeatingWater(model)
        heating_coil.setName(f"{self.SYSTEM_PREFIX} Central Heating Coil")

        fan = openstudio.epmodel.FanVariableVolume(model)
        fan.setName(f"{self.SYSTEM_PREFIX} Supply Fan")

        supply_outlet_node = air_loop.supplyOutletNode()
        for component in [outdoor_air_system, cooling_coil, heating_coil, fan]:
            if not self.require_condition(
                runner, component.addToNode(supply_outlet_node), f"Could not add {component.nameString()} to the VAV supply path."
            ):
                return None

        if not self.require_condition(
            runner,
            chilled_water_loop.addDemandBranchForComponent(cooling_coil)
            and hot_water_loop.addDemandBranchForComponent(heating_coil),
            "Could not connect the central water coils to their plant loops.",
        ):
            return None

        for name, node in [
            ("Deck Air", supply_outlet_node),
            ("Cooling Coil Outlet", heating_coil.airInletModelObject().get().to_Node().get()),
            ("Heating Coil Outlet", fan.inletModelObject().get().to_Node().get()),
        ]:
            spm = openstudio.epmodel.SetpointManagerScheduled(model)
            spm.setName(f"{self.SYSTEM_PREFIX} {name} Setpoint Manager")
            if not self.require_condition(
                runner,
                spm.setSchedule(deck_schedule) and spm.addToNode(node),
                f"Could not add the {name.lower()} setpoint manager.",
            ):
                return None

        return air_loop

    def run(self, model, runner, user_arguments):
        super().run(model, runner, user_arguments)
        if not runner.validateUserArguments(self.arguments(model), user_arguments):
            return False

        all_zones = sorted(model.getThermalZones(), key=lambda zone: zone.nameString())
        zones = [zone for zone in all_zones if not zone.thermostat().empty()]

        if not zones:
            runner.registerError("The model has no thermostatically controlled zones to condition.")
            return False

        existing = {
            "air loops": len(model.getAirLoopHVACs()),
            "zone HVAC components": len(model.getZoneHVACComponents()),
            "plant loops": len(model.getPlantLoops()),
            "ideal-loads systems": len(model.getZoneHVACIdealLoadsAirSystems()),
        }
        conflicts = [f"{count} {label}" for label, count in existing.items() if count]
        if conflicts:
            runner.registerError(
                "The multizone VAV system requires an HVAC-ready model, but found " + ", ".join(conflicts) + "."
            )
            return False

        pump_zone = zones[0]
        hot_water_loop = self.build_plant_loop(model, runner, pump_zone, True, len(zones))
        chilled_water_loop = self.build_plant_loop(model, runner, pump_zone, False, len(zones))

        if hot_water_loop is None or chilled_water_loop is None:
            runner.registerError("Could not link the plant pumps to a thermal zone for skin-loss gains.")
            return False

        hot_water_loop.setName(f"{self.SYSTEM_PREFIX} Hot Water Loop")
        chilled_water_loop.setName(f"{self.SYSTEM_PREFIX} Chilled Water Loop")

        air_loop = self.build_multizone_air_loop(model, runner, hot_water_loop, chilled_water_loop)
        if air_loop is None:
            return False
        for zone in zones:
            reheat_coil = openstudio.epmodel.CoilHeatingWater(model)
            reheat_coil.setName(f"{zone.nameString()} Reheat Coil")
            if not self.require_condition(
                runner,
                hot_water_loop.addDemandBranchForComponent(reheat_coil),
                f"Could not connect the reheat coil for '{zone.nameString()}'.",
            ):
                return False
            terminal = openstudio.epmodel.AirTerminalSingleDuctVAVReheat(
                model, model.alwaysOnDiscreteSchedule(), reheat_coil
            )
            terminal.setName(f"{zone.nameString()} VAV Reheat Terminal")
            if not self.require_condition(
                runner,
                air_loop.addBranchForZone(zone, terminal),
                f"Could not connect the VAV terminal for '{zone.nameString()}'.",
            ):
                return False

        if not self.require_condition(
            runner,
            len(air_loop.thermalZones()) == len(zones)
            and len(model.getAirTerminalSingleDuctVAVReheats()) == len(zones),
            "The VAV system does not serve every controlled zone exactly once.",
        ):
            return False

        runner.registerFinalCondition(
            f"Added one multizone VAV air system serving {len(zones)} controlled zone(s), with hot-water reheat, "
            f"central chilled water, and shared plant loops; left {len(all_zones) - len(zones)} unconditioned zone(s) without HVAC."
        )
        return True


EpModelAddMultizoneVAVSystem().registerWithApplication()
